### PR TITLE
Select barrier based on active plan

### PR DESCRIPTION
**Pull request by wenyuzhao**
Each plan configures its required barrier in `PlanConstraints`, and vm-bindings will enable the selected barrier if supported.

Note: currently only OpenJDK binding support mutator barriers.

https://github.com/mmtk/mmtk-openjdk/pull/44

# TODO

- [ ] Benchmark

### CODE REVIEW

**Comment by qinsoon** [comment]
openjdk-binding-test
OPENJDK_BINDING_REF=runtime-barrier-selection

**Comment by qinsoon** [comment]
openjdk-perf-compare
OPENJDK_BINDING_BRANCH_REF=runtime-barrier-selection

**Comment by qinsoon** [comment]
openjdk-microbm
OPENJDK_BINDING_BRANCH_REF=runtime-barrier-selection

File: src/plan/plan_constraints.rs
Comment (qinsoon) [inline src/plan/plan_constraints.rs:17]: Does this `barrier` refer to write barrier only? Should we merge `needs_write_barrier` with `barrier`? For example, remove `needs_write_barrier` - basically `BarrierSelector::NoBarrier` means `false` for `needs_write_barrier`.

Comment (qinsoon) [reply src/plan/plan_constraints.rs:17]: Should the `barrier` selector be consistent with the `barrier` field in `Mutator`? I would suggest adding a function `fn new_barrier(selector: BarrierSelector) -> Box<dyn Barrier>` to `barriers.rs` so that `Mutator.barrier` is created from the selector, and making constructors of the barrier implementations as private so that people can only create barrier from a selector. This could ensure the selector is always consistent with the actual barrier used.

**Review by qinsoon — APPROVED** [review]
Looks good to me. But I am concerned that we currently have a few fields about barrier, and we should try make them consistent. I suggest merging `PlanConstraints.needs_write_barrier` with `PlanConstraints.barrier`, and create `Mutator.barrier` from `PlanConstraints.barrier`.

File: src/plan/plan_constraints.rs
Comment (wenyuzhao) [reply src/plan/plan_constraints.rs:16]: I've removed `needs_write_barrier`.

For the second suggestion, I think it is hard to use a factory method for all Barrier struct creation. Different barrier type requires different constructor arguments. Some barriers require a specific type of nursery space while some barriers do not require any arguments (e.g. NoBarrier).

**Comment by javadamiri** [comment]
> Looks good to me. But I am concerned that we currently have a few fields about barrier, and we should try make them consistent. I suggest merging `PlanConstraints.needs_write_barrier` with `PlanConstraints.barrier`, and create `Mutator.barrier` from `PlanConstraints.barrier`.

Isn't it more informative to replace `barrier` with something like `barrier_type` or `barrier_selector`?

**Review by javadamiri — APPROVED** [review]
LGMT

**Comment by qinsoon** [comment]
We have two approvals for this PR. I think it is good to merge.
@caizixian Do you plan to review this?
@wenyuzhao Do you have anything you would like to address before merging?

**Comment by qinsoon** [comment]
> > Looks good to me. But I am concerned that we currently have a few fields about barrier, and we should try make them consistent. I suggest merging `PlanConstraints.needs_write_barrier` with `PlanConstraints.barrier`, and create `Mutator.barrier` from `PlanConstraints.barrier`.
> 
> Isn't it more informative to replace `barrier` with something like `barrier_type` or `barrier_selector`?

I think `barrier` is not ambiguous in the context. However, it is up to you and Wenyu.

**Comment by wenyuzhao** [comment]
> We have two approvals for this PR. I think it is good to merge.
> @caizixian Do you plan to review this?
> @wenyuzhao Do you have anything you would like to address before merging?

I don't have anything else to address

**Pull request merged**
